File: 261-graph-valid-tree/261-graph-valid-tree.scala
import scala.collection.mutable._

object Solution {
    def validTree(n: Int, edges: Array[Array[Int]]): Boolean = {
        val group: Array[Int] = Array.fill(n)(0)
        
        for(i <- 0 to n - 1) group(i) = i 
        
        def find(i: Int): Int = {
            if (group(i) != i) {
                val gid = find(group(i)) 
                group(i) = gid
                gid
            } else i
        }
        
        def union(a: Int, b: Int): Boolean = {
            val ga = find(a)
            val gb = find(b)
            
            if (ga == gb) false 
            else {
                group(ga) = gb
                true
            }
        }
        
        var hasLoop = false
        edges.foreach { edge => hasLoop ||= !union(edge(0), edge(1)) }
        
        !hasLoop && ((0 to n - 1).map(find(_)).toSet.size == 1)
    }
}

File: 261-graph-valid-tree/261-graph-valid-tree.java
class Solution {
    int[] gid = null;
    
    public boolean validTree(int n, int[][] edges) {
        gid = new int[n];
        
        for(int i = 0; i < n; i++) gid[i] = i;
        
        for(int[] edge : edges) {
            if (!join(edge[0], edge[1])) return false;
        }
        
        int g = find(0);
        for(int i = 1; i < n; i++) {
            if (find(i) != g) return false;
        }
        
        return true;
    }
    
    int find(int n) {
        if (gid[n] != n) {
            gid[n] = find(gid[n]);
            return gid[n];
        }
        return n;
    }
    
    boolean join(int a, int b) {
        int ga = find(a);
        int gb = find(b);
        if (ga == gb) return false;
        
        gid[gb] = ga;
        
        return true;
    }
}